File: set4/challenge_29.cpp
#include <iostream>
#include <string>
#include <random>
#include <cassert>
#include <algorithm>
#include <openssl/sha.h>

#include "cryptopp/cryptlib.h"
#include "cryptopp/filters.h"
#include "cryptopp/sha.h"
#include "cryptopp/hex.h"

std::string key;

bool validate_mac(std::string &input, std::string &key, std::string &mac) {
    std::string key_prefixed_message = key + input;
    std::string hash;

    CryptoPP::SHA1 sha1;
    CryptoPP::StringSource ss1(key_prefixed_message, true,
            new CryptoPP::HashFilter(sha1,
                new CryptoPP::HexEncoder(
                    new CryptoPP::StringSink(hash)
                )
            )
    );

    return (mac == hash) ? true : false;
}

std::string md_pad(std::string &input) {
    if (input.length() % 64 == 0) {
        return input;
    }

    unsigned int num_full_blocks = input.length() / 64;
    unsigned int remaining_bytes = input.length() - (num_full_blocks * 64);

    std::string last_block = input.substr(num_full_blocks * 64, remaining_bytes);

    unsigned int num_zeroes = ((64 - (((last_block.length() + 1 + 8) % 64))) % 64);

    // Append a character that has a '1' in the MSB
    last_block += '\x80';

    // Append the zeroes
    last_block.append(num_zeroes, '\x00');

    unsigned long long length = remaining_bytes * 8;

    // Append the size of the message, THIS ONLY WORKS ON LITTLE ENDIAN MACHINES
    std::string length_string = "";
    length_string.append(reinterpret_cast<char*>(&length), sizeof(length));
    std::reverse(length_string.begin(), length_string.end());
    last_block += length_string;

    std::string output = input.substr(0, num_full_blocks * 64) + last_block;

    return output;
}

void generate_key(std::string &input) {
    auto seed = std::random_device()();
    std::mt19937 rng(seed);
    
    key = input.substr(0, (rng() % input.length()));
}

std::string sha1_hash(std::string &input, std::string &key) {
    std::string key_mac;

    CryptoPP::SHA1 sha1;
    CryptoPP::StringSource ss1(key + input, true,
            new CryptoPP::HashFilter(sha1,
                new CryptoPP::HexEncoder(
                    new CryptoPP::StringSink(key_mac)
                )
            )
    );

    return key_mac;
}

void sha1_length_extension_attack(std::string &mac, std::string &message, std::string &extension) {

}

int main(void) {
    std::string original_message {"comment1=cooking%20MCs;userdata=foo;comment2=%20like%20a%20pound%20of%20bacon"};
    generate_key(original_message);
    std::string message_mac = sha1_hash(original_message, key);
    std::string extension = ";admin=true";
    sha1_length_extension_attack(message_mac, original_message, extension);
}

File: set4/challenge_29.py
#!/usr/bin/env python2

import os
import sha1
import struct

key = os.urandom(16)

def validate_mac(message, key, mac):
    output_mac = sha1.Sha1Hash().update(key + message).hexdigest()
    if output_mac == mac:
        return True
    else:
        return False 

def sha1_pad(message):
    padding = ""

    # Add a byte with 1 in the MSB
    padding += "\x80"

    # Pad the string until it is a multiple of 64 bytes with 2 words open
    padding += '\x00' * ((56 - (len(message) + 1) % 64) % 64)

    # Append the length of the bitstring in number of bits, big endian style
    padding += struct.pack(">Q", (len(message) * 8))

    return message + padding

def get_registers(mac):
    return struct.unpack(">IIIII", mac)

def sha1_length_attack(message, mac, extension):
    registers = list(get_registers(mac))

    for i in xrange(0, 40):
        padded_text = sha1_pad(("A" * i) + message)
        forged_mac = sha1.Sha1Hash(h=registers, message_byte_length=len(padded_text)).update(extension).hexdigest()
        extended_message = padded_text[i:] + extension
        if validate_mac(extended_message, key, forged_mac) is True:
            return i, forged_mac

    return -1, ""

if __name__ == "__main__":
    message = "comment1=cooking%20MCs;userdata=foo;comment2=%20like%20a%20pound%20of%20bacon"
    mac = sha1.Sha1Hash().update(key + message).digest()

    key_len, forged_mac = sha1_length_attack(message, mac, ";admin=true")

    print("The key length is {} with a forged MAC of {}".format(key_len, forged_mac))
